Linux/sgx/test_app/enclave/tests/evp_smx.c

```diff
@@ -0,0 +1,620 @@
+
+    do {
+        // 1. Create a new EC key
+        ec_key = EC_KEY_new();
+        if (ec_key == NULL) {
+            printf("Error: fail to create a new EC key\n");
+            ret = -1;
+            break;
+        }
+
+        // 2. Set the new EC key's curve
+        if (EC_KEY_set_group(ec_key, ec_group) != 1) {
+            printf("Error: fail to set the new EC key's curve\n");
+            ret = -2;
+            break;
+        }
+
+        // 3. Generate key pair based on the curve
+        if (!EC_KEY_generate_key(ec_key)) {
+            printf("Error: fail to generate key pair based on the curve\n");
+            ret = -3;
+            break;
+        }
+
+        // 4. Generate SM2 private key based on the curve
+        pri_bio = BIO_new(BIO_s_mem());
+        if (pri_bio == NULL) {
+            printf("Error: fail to create a BIO for private key\n");
+            ret = -4;
+            break;			
+        }
+        if (!PEM_write_bio_ECPrivateKey(pri_bio, ec_key, NULL, NULL, 0, NULL, NULL)) {
+            printf("Error: fail to write private key from ec_key to the BIO\n");
+            ret = -5;
+            break;			
+        }
+        pri_len = BIO_pending(pri_bio);
+        if (pri_len == 0) {
+            printf("Error: fail to get size of the BIO for private key\n");
+            ret = -6;
+            break;
+        }
+        *private_key = (char*)malloc(pri_len+1);
+        if (BIO_read(pri_bio, *private_key, pri_len) <= 0) {
+            printf("Error: fail to read private key from the BIO\n");
+            ret = -7;
+            break;
+        }
+        (*private_key)[pri_len] = '\0';
+
+        // 5. Generate SM2 public key based on the curve
+        pub_bio = BIO_new(BIO_s_mem());
+        if (pub_bio == NULL) {
+            printf("Error: fail to create a BIO for public key\n");
+            ret = -8;
+            break;
+        }
+        if (!PEM_write_bio_EC_PUBKEY(pub_bio, ec_key)) {
+            printf("Error: fail to write public key from ec_key to the BIO\n");
+            ret = -9;
+            break;
+        }
+        pub_len = BIO_pending(pub_bio);
+        if (pub_len == 0) {
+            printf("Error: fail to get size of the BIO for public key\n");
+            ret = -10;
+            break;
+        }
+        *public_key = (char*)malloc(pub_len+1);
+        if (BIO_read(pub_bio, *public_key, pub_len) <= 0) {
+            printf("Error: fail to read public key from the BIO\n");
+            ret = -11;
+            break;
+        }
+        (*public_key)[pub_len] = '\0';
+
+    } while(0);
+
+    // 6. Finalize
+    if (ret != 0)
+    {
+        SAFE_FREE(*private_key, pri_len+1);
+        SAFE_FREE(*public_key, pub_len+1);
+    }	
+	EC_KEY_free(ec_key);
+	BIO_free_all(pri_bio);
+	BIO_free_all(pub_bio);
+
+    return ret;
+}
+
+// sign the message
+static int sign_sm2(const char* private_key, char* data, size_t data_size, unsigned char** signature, size_t* sign_len)
```

Comment:
Suggest to change the type of signature to char*, we don't allocate buffer inside sign_sm2(), so no need to use char **. It might confuse the users.